FinderCompass/src/main/java/atomicstryker/findercompass/client/ThreadCompassWorker.java

```diff
@@ -1 +1 @@
		}
		
		isRunning = false;
	}
	
    private ChunkCoordinates findNearestBlockChunkOfIDInRange(Set<BlockData> blocks, int playerX, int playerY, int playerZ, int xzRange, int yRange, int minY, int maxY)
    {
        List<ChunkCoordinates> blocksInRange = this.findBlocksOfIDInRange(blocks, playerX, playerY, playerZ, xzRange, yRange, minY, maxY);
        ChunkCoordinates playerCoords = new ChunkCoordinates(playerX, playerY, playerZ);
        ChunkCoordinates resultCoords = new ChunkCoordinates(0, 0, 0);
        double minDist = 9999.0D;

        for (int i = 0; i < blocksInRange.size(); ++i)
        {
            ChunkCoordinates coords = (ChunkCoordinates)blocksInRange.get(i);
            double localDist = this.getDistanceBetweenChunks(playerCoords, coords);
            if (localDist < minDist)
            {
                resultCoords = coords;
                minDist = localDist;
            }
        }
        
        //System.out.printf("Compassworker found stuff of id %s at [%d|%d|%d]\n", blockID, resultCoords.posX, resultCoords.posY, resultCoords.posZ);
        return resultCoords;
    }

    private List<ChunkCoordinates> findBlocksOfIDInRange(Set<BlockData> blocks, int playerX, int playerY, int playerZ, int xzRange, int yRange, int minY, int maxY)
    {
        ArrayList<ChunkCoordinates> resultList = new ArrayList<ChunkCoordinates>();

        for (int yIter = playerY - yRange - 1; yIter <= playerY + yRange; ++yIter)
        {
            if (yIter >= minY && yIter <= maxY)
            {
                for (int zIter = playerZ - xzRange; zIter <= playerZ + xzRange; ++zIter)
                {
                    for (int xIter = playerX - xzRange; xIter <= playerX + xzRange; ++xIter)
                    {
                    	for (BlockData blockData : blocks)
                    	{
                    		if (this.mcinstance.theWorld.getBlock(xIter, yIter, zIter) == blockData.getBlockID())
                    		{
                    			if (blockData.getDamage() > 0 && mcinstance.theWorld.getBlockMetadata(xIter, yIter, zIter) != blockData.getDamage())
                    			{
                    				continue;
                    			}
                            
                    			ChunkCoordinates var13 = new ChunkCoordinates(xIter, yIter, zIter);
                    			resultList.add(var13);
                    		}
                    		Thread.yield();
                    	}
                    }
                }
            }
        }

        return resultList;
    }
    
    private double getDistanceBetweenChunks(ChunkCoordinates var1, ChunkCoordinates var2)
    {
        int xdiff = Math.abs(var1.posX - var2.posX);
        int ydiff = Math.abs(var1.posY - var2.posY);
        int zdiff = Math.abs(var1.posZ - var2.posZ);
        return Math.sqrt(Math.pow((double)xdiff, 2.0D) + Math.pow((double)ydiff, 2.0D) + Math.pow((double)zdiff, 2.0D));
    }
}

```

Comment:
My email client displays this correctly, for whatever reason. So i've checked out your critical "inner loop" and it is a bit inefficient - instead of storing the world.getBlock result and comparing it against the blocklist, you call getBlock once for every blocklist entry. Let's assume you have something like mod copper with 5 types, that is 4 superfluous getBlock calls per block, in a 16 by 16 area, 4 blocks height, that's a measly 4096 getBlock calls that could have been avoided. Please amend this.